shell/browser/api/electron_api_system_preferences_mac.mm

```diff
@@ -121,6 +121,121 @@ AVMediaType ParseMediaType(const std::string& media_type) {
   }
 }
 
+v8::Local<v8::Value> GetUserDefaultImpl(v8::Isolate* isolate,
+                                        NSUserDefaults* defaults,
+                                        const std::string& name,
+                                        const std::string& type) {
+  NSString* key = base::SysUTF8ToNSString(name);
+  if (type == "string") {
+    return gin::StringToV8(
+        isolate, base::SysNSStringToUTF8([defaults stringForKey:key]));
+  } else if (type == "boolean") {
+    return v8::Boolean::New(isolate, [defaults boolForKey:key]);
+  } else if (type == "float") {
+    return v8::Number::New(isolate, [defaults floatForKey:key]);
+  } else if (type == "integer") {
+    return v8::Integer::New(isolate, [defaults integerForKey:key]);
+  } else if (type == "double") {
+    return v8::Number::New(isolate, [defaults doubleForKey:key]);
+  } else if (type == "url") {
+    return gin::ConvertToV8(isolate,
+                            net::GURLWithNSURL([defaults URLForKey:key]));
+  } else if (type == "array") {
+    return gin::ConvertToV8(isolate,
+                            NSArrayToListValue([defaults arrayForKey:key]));
+  } else if (type == "dictionary") {
+    return gin::ConvertToV8(isolate, NSDictionaryToDictionaryValue(
+                                         [defaults dictionaryForKey:key]));
+  } else {
+    return v8::Undefined(isolate);
+  }
+}
+
+void SetUserDefaultImpl(NSUserDefaults* defaults,
+                        const std::string& name,
+                        const std::string& type,
+                        gin_helper::Arguments* args) {
+  const auto throwConversionError = [&] {
+    args->ThrowError("Unable to convert value to: " + type);
+  };
+
+  NSString* key = base::SysUTF8ToNSString(name);
+  if (type == "string") {
+    std::string value;
+    if (!args->GetNext(&value)) {
+      throwConversionError();
```

Comment:
I think we can refactor the code flow to avoid duplicating the `throwConversionError` call.

```c++
if (type == "string") {
  std::string value;
  if (args->GetNext(&value)) {
    [defaults setObject:base::SysUTF8ToNSString(value) forKey:key];
    return;
  }
} else if (...) {
  ...
} else {
  args->ThrowError("Invalid type: " + type);
  return;
}

args->ThrowError("Unable to convert value to: " + type);
```

Reply:
I just moved the function as is with minimal changes to be able to re-use it. I can refactor it in a follow-up PR.

Reply:
actually, let's do it now